modules/library/coverage/src/main/java/org/geotools/coverage/processing/operation/Resampler2D.java

```diff
@@ -493,373 +853,235 @@ public static GridCoverage2D reproject(
-        }
-        targetHints.put(JAI.KEY_REPLACE_INDEX_COLOR_MODEL, false);
-        targetHints.put(JAI.KEY_TRANSFORM_ON_COLORMAP, false);
+    private static ResampleState resolveParameters(
+            final GridCoverage2D providedSourceCoverage,
+            final CoordinateReferenceSystem providedTargetCRS,
+            final GridGeometry2D providedTargetGG,
+            final Interpolation providedInterpolation,
+            final Hints providedHints,
+            double[] providedBackgroundValues)
+            throws FactoryException {
 
-        if (hints != null) {
-            targetHints.add(hints);
-        }
-        ImageLayout layout = (ImageLayout) targetHints.get(JAI.KEY_IMAGE_LAYOUT);
-        if (layout != null) {
-            layout = (ImageLayout) layout.clone();
-        } else {
-            layout = new ImageLayout();
-            // Do not inherit the color model and sample model from the 'sourceImage';
-            // Let the operation decide itself. This is necessary in case we change the
-            // source, as we do if we choose the "Mosaic" operation.
-        }
-        final Rectangle sourceBB = sourceGG.getGridRange2D();
-        final Rectangle targetBB = targetGG.getGridRange2D();
-        if (isBoundsUndefined(layout, false)) {
-            layout.setMinX(targetBB.x);
-            layout.setMinY(targetBB.y);
-            layout.setWidth(targetBB.width);
-            layout.setHeight(targetBB.height);
+        Utilities.ensureNonNull("sourceCoverage", providedSourceCoverage);
+
+        ResampleState state = new ResampleState(providedSourceCoverage);
+        state.hints = providedHints == null ? new Hints() : new Hints(providedHints);
+        state.sourceCRS = state.sourceCoverage.getCoordinateReferenceSystem();
+        state.targetCRS = resolveTargetCRS(state.sourceCRS, providedTargetCRS, providedTargetGG);
+        state.interpolation = prepareHints(providedInterpolation, state.hints);
+        state.backgroundValues = providedBackgroundValues;
+
+        state.sourceGridGeometry = state.sourceCoverage.getGridGeometry();
+        state.automaticGridRange = shallComputeGridRange(providedTargetGG);
+        state.automaticGridGeometry =
+                shallComputeGridGeometry(providedTargetGG, state.automaticGridRange);
+        state.targetGridGeometry =
+                computeTargetGridGeometry(
+                        state.sourceCoverage, providedTargetGG, state.automaticGridRange);
+
+        @SuppressWarnings("unchecked")
+        Map<String, Object> sourceProps = state.sourceCoverage.getProperties();
+        state.sourceProps = sourceProps != null ? new HashMap<>(sourceProps) : new HashMap<>();
+
+        return state;
+    }
+
+    private static ResampleState resolveRoiAndNoData(ResampleState state) {
+        /*
+         * The projection are usually applied on floating-point values, in order
```

Comment:
Hm... this comment seems to be inappropriate here, but it was kind of dangling in the original code too, I cannot find the actual code bit where it applies...